File: twitch_solutions/easy/is_balanced.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def isLeaf(self, node):
        return node.left is None and node.right is None

    def length(self, node, ln):
        if self.isLeaf(node):
            return ln
        len_left = ln
        if node.left:
            len_left = self.length(node.left, ln + 1)
        len_right = ln
        if node.right:
            len_right = self.length(node.right, ln + 1)
        return 



    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        return abs(self.length(root.left, 0) - self.length(root.right, 0)) < 2

File: twitch_solutions/go/easy/is_balanced.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type NodeNLvl struct {
	Node   *TreeNode
	Height int
}

var heightMap map[*TreeNode]int

func cleanUp() {
	heightMap = make(map[*TreeNode]int)
}

func height(node *TreeNode) int {
	if node == nil {
		return 0
	}
	if val, ok := heightMap[node]; ok {
		return val
	}
	height := 1 + int(math.Max(float64(height(node.Left)), float64(height(node.Right))))
	heightMap[node] = height
	return height
}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		cleanUp()
		return true
	}
	if int(math.Abs(float64(height(root.Left)-height(root.Right)))) > 1 {
		cleanUp()
		return false
	}
	if root.Left != nil {
		if !isBalanced(root.Left) {
			cleanUp()
			return false
		}
	}
	if root.Right != nil {
		if !isBalanced(root.Right) {
			cleanUp()
			return false
		}
	}

	cleanUp()
	return true
}
